### ISSUE TITLE
Support notify hook for thread resume and suspend

### ISSUE THREAD

**Issue by greglink** [eclipse-threadx/threadx#426]
**Problem Description**
I'd like to monitor/observe (at runtime) the latency from a thread becoming ready for any reason (timer, semaphore, etc) until it begins execution, and then voluntarily suspends itself. These metrics are useful for ensuring a responsive real time operating system, and can be emitted as telemetry in formats like a histogram or high water mark to indicate potential delays or timing issues.

**Describe the solution you'd like**
I'd like to add/install optional callback/notify hooks in a similar fashion to the [tx_thread_entry_exit_notify_](https://github.com/eclipse-threadx/threadx/blob/485a02faec6edccef14812ddce6844af1d7d2eef/common/src/tx_thread_entry_exit_notify.c#L76) hooks, only this time for suspend and resume.

**Describe alternatives you've considered**
I could theoretically instrument all potential resume initiators (e.g. queues, semaphores, ...) but it seems like instrumenting at tx_thread_resume or tx_thread_system_resume might be the best/simplest location. I also considered emitting trace information, and scanning the trace information at runtime, but this seemed very heavy, as I couldn't restrict trace emission to only those events.

**Additional context**
I'm volunteering to write the PR, if someone can re-assure me that this won't be a rejected request for performance reasons, and inform me of whether [_tx_thread_resume](https://github.com/eclipse-threadx/threadx/blob/485a02faec6edccef14812ddce6844af1d7d2eef/common_smp/src/tx_thread_resume.c#L74), [_tx_thread_system_resume](https://github.com/eclipse-threadx/threadx/blob/485a02faec6edccef14812ddce6844af1d7d2eef/common/src/tx_thread_system_resume.c#L80), or some other location is the best place to install the hook.

**Comment by fdesbiens** [comment]
Hi @greglink.

Thank you for opening this issue. I appreciate your offer to implement this. Let's see what the project team thinks of your idea.

@eclipse-threadx/iot-threadx-committers Please review the proposal and provide your feedback.

**Comment by billlamiework** [comment]
Most developers would not find this helpful, so anything added must be under conditional compilation. This is also a critical real-time code path, so we must be extra cautious. That said, perhaps extension macros (defaulted to white space but able to be overridden by the developer) might be safer than function pointers and result in less overhead. You could define these as function calls or directly toggle pins from inside these routines. Interested to hear your thoughts....

**Comment by billlamiework** [comment]
I agree Greg that the entry/exit function is more overhead-wise since it happens at every context switch. However, the historical reason for the function pointer in this case was because not all developers had access to the source code for rebuilding/overriding. That is moot now given the source is in the public domain.  Hence, my recommendation for a couple of extension macros....

**Comment by billlamiework** [comment]
I suggest a couple macros in tx_api.h defined to whitespace:

#ifndef TX_THREAD_SYSTEM_RESUME_EXTENSION
#define TX_THREAD_SYSTEM_RESUME_EXTENSION
#endif

#ifndef TX_THREAD_SYSTEM_SUSPEND_EXTENSION
#define TX_THREAD_SYSTEM_SUSPEND_EXTENSION
#endif

And then simply drop in TX_THREAD_SYSTEM_RESUME_EXTENSION and TX_THREAD_SYSTEM_SUSPEND_EXTENSION in the actual C code where you want to add the instrumentation. For resumption, I recommend sometime after the thread is completely on the ready list and for suspension sometime after the thread is completely removed from the ready list. 

I hope that helps!
